src/coreclr/src/vm/appdomain.hpp

```diff
@@ -1845,6 +1845,152 @@ class AppDomain : public BaseDomain
+            CrstHolder ch(pAppDomain->GetAssemblyListLock());
+            return GetCount_Unlocked();
+        }
+
+        DWORD GetCount_Unlocked()
+        {
+            CONTRACTL {
+                NOTHROW;
+                GC_NOTRIGGER;
+                MODE_ANY;
+            } CONTRACTL_END;
+
+#ifndef DACCESS_COMPILE
+            _ASSERTE(dbg_m_pAppDomain->GetAssemblyListLock()->OwnedByCurrentThread());
+#endif
+            // code:Append_Unlock guarantees that we do not have more than MAXDWORD items
+            return m_array.GetCount();
+        }
+
+        // Doesn't lock the assembly list (caller has to hold the lock already).
+        DomainCompositeImage* Get_UnlockedNoReference(DWORD index)
+        {
+            CONTRACTL {
+                NOTHROW;
+                GC_NOTRIGGER;
+                MODE_ANY;
+                SUPPORTS_DAC;
+            } CONTRACTL_END;
+
+#ifndef DACCESS_COMPILE
+            _ASSERTE(dbg_m_pAppDomain->GetAssemblyListLock()->OwnedByCurrentThread());
+#endif
+            return dac_cast<PTR_DomainCompositeImage>(m_array.Get(index));
+        }
+
+#ifndef DACCESS_COMPILE
+        void Set(AppDomain* pAppDomain, DWORD index, DomainCompositeImage* pCompositeImage)
+        {
+            CONTRACTL {
+                NOTHROW;
+                WRAPPER(GC_TRIGGERS); // Triggers only in MODE_COOPERATIVE (by taking the lock)
+                MODE_ANY;
+            } CONTRACTL_END;
+
+            _ASSERTE(dbg_m_pAppDomain == pAppDomain);
+
+            CrstHolder ch(pAppDomain->GetAssemblyListLock());
+            return Set_Unlocked(index, pCompositeImage);
+        }
+
+        void Set_Unlocked(DWORD index, DomainCompositeImage* pCompositeImage)
+        {
+            CONTRACTL {
+                NOTHROW;
+                GC_NOTRIGGER;
+                MODE_ANY;
+            } CONTRACTL_END;
+
+            _ASSERTE(dbg_m_pAppDomain->GetAssemblyListLock()->OwnedByCurrentThread());
+            m_array.Set(index, pCompositeImage);
+        }
+
+        HRESULT Append_Unlocked(DomainCompositeImage* pCompositeImage)
+        {
+            CONTRACTL {
+                NOTHROW;
+                GC_NOTRIGGER;
+                MODE_ANY;
+            } CONTRACTL_END;
+
+            _ASSERTE(dbg_m_pAppDomain->GetAssemblyListLock()->OwnedByCurrentThread());
+            return m_array.Append(pCompositeImage);
+        }
+
+#else //DACCESS_COMPILE
+        void
+        EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
+        {
+            SUPPORTS_DAC;
+
+            m_array.EnumMemoryRegions(flags);
+        }
+#endif // DACCESS_COMPILE
+
+        // Should be used only by code:AssemblyIterator::Create
+        ArrayList::Iterator GetArrayListIterator()
+        {
+            return m_array.Iterate();
+        }
+    };  // class DomainAssemblyList
+
+    // Conceptually a list of DomainCompositeImage structures, protected by lock AppDomain::GetAssemblyListLock
+    DomainCompositeImageList m_CompositeImages;
```

Comment:
It may be better for this list to live in AssemblyLoadContext (https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/assemblyloadcontext.h)

Reply:
Thanks for the suggestion, moved in 7th commit.